state/indexer/block/kv/util.go

```diff
@@ -108,24 +107,38 @@ func parseHeightFromEventKey(key []byte) (int64, error) {
 
 func parseEventSeqFromEventKey(key []byte) (int64, error) {
 	var (
-		compositeKey, typ, eventValue string
-		height                        int64
-		eventSeq                      int64
+		compositeKey, eventValue string
+		height                   int64
+		eventSeq                 int64
 	)
 
-	remaining, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &typ)
+	remaining, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height)
 	if err != nil {
 		return 0, fmt.Errorf("failed to parse event key: %w", err)
 	}
 
-	// This is done to support previous versions that did not have event sequence in their key
+	// We either have an event sequence or a function type (potentially) followed by an event sequence.
+	// Potential scenarios:
+	// 1. Events indexed with v0.38.x and later, will only have an event sequence
+	// 2. Events indexed between v0.34.27 and v0.37.x will have a function type and an event sequence
+	// 3. Events indexed before v0.34.27 will only have a function type
+	// function_type = 'being_block_event' | 'end_block_event'
 	if len(remaining) != 0 {
-		remaining, err = orderedcode.Parse(remaining, &eventSeq)
-		if err != nil {
-			return 0, fmt.Errorf("failed to parse event key: %w", err)
-		}
-		if len(remaining) != 0 {
-			return 0, fmt.Errorf("unexpected remainder in key: %s", remaining)
+		var typ string
+		remaining2, err := orderedcode.Parse(remaining, &typ) // Check if we have scenarios 2. or 3. (described above).
+		if err != nil {                                       // If it cannot parse the event function type, it could be 1.
+			remaining, err2 := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &eventSeq)
+			if err2 != nil || len(remaining) != 0 { // We should not have anything else after the eventSeq.
+				return 0, fmt.Errorf("failed to parse event key: %w; and %w", err, err2)
+			}
+
+		} else {
+			remaining, err2 := orderedcode.Parse(remaining2, &eventSeq) // If the event follows the scenario in 2.,
+			// retrieve the eventSeq, otherwise, there should be no remainder
+			// and no error and this parsing has no effect.
+			if err2 != nil || len(remaining) != 0 { // We should not have anything else after the eventSeq if in 2.
+				return 0, fmt.Errorf("failed to parse event key: %w", err2)
+			}
```

Comment:
```suggestion
			if len(remaining2) != 0 { // Are we in case 2 or 3
				remaining, err2 := orderedcode.Parse(remaining2, &eventSeq) // the event follows the scenario in 2.,
				// retrieve the eventSeq
				// there should be no error
				if err2 != nil { // We should not have anything else after the eventSeq if in 2.
					return 0, fmt.Errorf("failed to parse event key: %w", err2)
				}
			}
```

I'm suggesting this, because I don't clearly see the path taken by scenario 3 in the present code.
I also see two main advantages with my suggestion:
* Easier to distinguish paths followed by scenarios 2 and 3
* You don't call `Parse` unless you're sure you need to call it (cleaner)

Reply:
I am fine with the chaneg but we still need to check whether len(remaining) != 0 because that is invalid. I will push a change including both. Let me know if you agree.